### DISCUSSION TITLE
Trouble caused by cached pages in Safari

### DISCUSSION THREAD

**Discussion by AsteroidOrangeJuice** [authelia/authelia#7510] [Troubleshooting] [answered]
I have multiple applications running behind NGINX using Authelia for Auth. Occasionally, I hit an issue where my main webpage loads fine, but the assets loaded with the main page (the css and js) do not.

Upon deeper investigation, it appears that when this happens when both:

1. I'm loading a cached version of the webpage itself, but the JS/CSS is not cached. 
2. I'm navigating to my site from an unrelated domain.

When fetching the CSS/JS on the initial page load in this state, it is not including the appropriate token for authelia to see I'm auth'd. I can actually see when it tries to load that JS/CSS file, instead of returning the file I'd expect, I get the home page for authelia, which is obviously not desirable in this scenario.

This issue happens regardless of my backend, I primarily have two type of applications - one is a simple Flask webserver with basic HTML, JS, & CSS. The other is a static website built with https://github.com/rust-lang/mdBook.

This is happening on Mac Safari Version 17.5 (19618.2.12.11.6).

Based on https://github.com/authelia/authelia/discussions/6106 I found this issue which is probably the technical root cause: https://bugs.webkit.org/show_bug.cgi?id=213510

> "We found a bug in our app (using WKWebView) on iOS 14 developer beta that images that are loaded from another origin don't load because the necessary cookie is not sent."

However, this bug has been around on Safari since 2020. Is there anything we can do to make this either succeed, or fail more gracefully? I can't tell all my customers to simply not use Safari, unfortunately.

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

The issue you're experiencing with Safari not sending the necessary cookies for assets like CSS and JS when navigating from an unrelated domain or loading a cached version of the webpage is indeed related to the WebKit bug you mentioned. This bug affects how cookies are handled in cross-origin requests, which is crucial for Authelia's authentication mechanism.

Here are a few potential solutions or workarounds you can consider:

1. **SameSite Cookie Attribute**: Ensure that the cookies set by Authelia have the `SameSite=None` attribute and are marked as `Secure`. This can help in making sure that cookies are sent with cross-origin requests. You can configure this in your Authelia configuration.

2. **Service Workers**: Implementing a service worker can help in caching the assets more effectively and ensuring that the requests for these assets include the necessary cookies. This might require some development effort but can provide a more robust solution.

3. **Custom Error Handling**: Modify your NGINX configuration to handle the scenario where the assets are not loaded correctly. You can set up custom error pages or redirects that inform the user to refresh the page or try a different browser.

4. **Cache-Control Headers**: Adjust the cache-control headers for your assets to ensure they are cached along with the main page. This can help in reducing the frequency of the issue.

5. **User Agent Detection**: As a last resort, you can detect if the user is using Safari and provide a warning or alternative instructions. This is not ideal but can help in mitigating the issue for affected users.

Here is an example of how you might configure the `SameSite` attribute in your Authelia configuration:

```yaml
session:
  name: authelia_session
  secret: your_secret
  expiration: 3600
  inactivity: 300
  remember_me_duration: 1M
  domain: example.com
  same_site: None
  secure: true
```

For more detailed configuration, you can refer to the [Authelia documentation](https://www.authelia.com/integration/prologue/get-started).

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by AsteroidOrangeJuice** [comment]
Also, happy to provide whatever other information is useful. I can upload a HAR if there's a way to upload them privately.

**Comment by AsteroidOrangeJuice** [comment]
Also, SameSite=None no longer works due to https://stackoverflow.com/questions/58525719/safari-not-sending-cookie-even-after-setting-samesite-none-secure

> Safari is no longer sending cross-site cookies. So even if the cookie is set to SameSite=None it will not be sent with the third party ajax requests.

Again, there's a workaround, but I can't expect everyone to change to more insecure settings just for this usecase.

**Comment by james-d-elliott** [comment, marked as answer]
I suspect the only fix is to configure all responses with headers that force Safari to not cache anything other than cachable content (protected primary URLs are **_not_** cachable in this scenario). As to what that combination of headers is for Safari I'm not sure; as you've seen it operates completely at odds with every other browser.

What's worse is there is no longer a free way to test it as you must own their hardware. 

Try:

```
Cache-Control: no-store
Vary: *
```

**Comment by AsteroidOrangeJuice** [comment]
Documenting what worked for me:

I added this to my authelia.conf:

```
    # stop caching bad js and css files
    add_header Cache-Control 'no-store, no-cache' always;
```

And I added this to my authelia-authrequest.conf:

```
# only redirect if a 401 and it's not a css or js file
# allow the 401 to propagate on css and js
if ($request_uri !~ "(\.css|\.js)$") {
    error_page 401 =302 https://authelia.mydomain.com/?rd=$target_url;
}
```

I think the first one helped, but the second one is the one which really solved the problem. Both work well in combination though, I have had 0 errors with this since ~early July.
